### PR TITLE
EOS-26323: interface to allocate and return a new fid in hax

### PR DESCRIPTION
**Pull request by mssawant**
For dtm to understand if a motr client logs must be discarded, everytime
a motr client restarts, a new fid must be allocated by Hare when Hare
receives an entrypoint request from that motr client.

Solution:
- Hare reserves 32 bits from motr object fid to be used at runtime.
- On motr client restart, when hax receives an entrypoint request,
  hax increments the corresponding motr configuration object type
  sequence and sets it in the reserved 32 bit of the motr process
  configuration object fid.
- Updated fid is saved the Consul corresponding to its base fid.
- Updated motr client process fid is piggy backed over the outgoing
  ha messages for a given process from hax to other motr processes.

Signed-off-by: Mandar Sawant <mandar.sawant@seagate.com>

### CODE REVIEW

**Comment by mssawant** [comment]
```
[root@ssc-vm-1623:root] hctl status
Data pool:
    # fid name
    0x6f00000000000001:0x1 'sns-pool'
Profile:
    # fid name: pool(s)
    0x7000000000000001:0x0 'default': 'sns-pool' 'dix-pool' None
Services:
    ssc-vm-1623.colo.seagate.com  (RC)
    [started]  hax        0x7200000000000001:0x0   inet:tcp:10.230.250.51@2001
    [started]  confd      0x7200000000000001:0x1   inet:tcp:10.230.250.51@3001
    [started]  ioservice  0x7200000000000001:0x2   inet:tcp:10.230.250.51@3002
    [started]  ioservice  0x7200000000000001:0x3   inet:tcp:10.230.250.51@3003
    [started]  s3server   0x7200000000000001:0x4   inet:tcp:10.230.250.51@4001
    [offline]  m0_client  0x7200000000000001:0x5   inet:tcp:10.230.250.51@5001
    [unknown]  m0_client  0x7200000000000001:0x6   inet:tcp:10.230.250.51@5002
[root@ssc-vm-1623:root]
ssc-vm-1623.colo.seagate.com/processes/0x7200000000000001:0x0:{"state": "M0_CONF_HA_PROCESS_STARTED", "type": "M0_CONF_HA_PROCESS_M0D"}
ssc-vm-1623.colo.seagate.com/processes/0x7200000000000001:0x1:{"state": "M0_CONF_HA_PROCESS_STARTED", "type": "M0_CONF_HA_PROCESS_M0D"}
ssc-vm-1623.colo.seagate.com/processes/0x7200000000000001:0x2:{"state": "M0_CONF_HA_PROCESS_STARTED", "type": "M0_CONF_HA_PROCESS_M0D"}
ssc-vm-1623.colo.seagate.com/processes/0x7200000000000001:0x3:{"state": "M0_CONF_HA_PROCESS_STARTED", "type": "M0_CONF_HA_PROCESS_M0D"}
ssc-vm-1623.colo.seagate.com/processes/0x7200000000000001:0x4:{"state": "M0_CONF_HA_PROCESS_STARTED", "type": "M0_CONF_HA_PROCESS_OTHER"}
ssc-vm-1623.colo.seagate.com/processes/0x7200000000000001:0x5:{"state": "M0_CONF_HA_PROCESS_STOPPED", "type": "M0_CONF_HA_PROCESS_OTHER"}
stats/filesystem:{"stats": {"fs_free_seg": 269928096, "fs_total_seg": 270469216, "fs_free_disk": 161061273600, "fs_avail_disk": 161061273600, "fs_total_disk": 161061273600, "fs_svc_total": 4, "fs_svc_replied": 4}, "timestamp": 1641872219.845349, "date": "2022-01-10T20:36:59.845349"}
[root@ssc-vm-1623:root] m0client -l 'inet:tcp:10.230.250.51@5001' -H 'inet:tcp:10.230.250.51@2001' -p '<0x7000000000000001:0x0>' -P '0x7200000000000001:0x5'
m0client >>quit
Help:
touch   OBJ_ID
write   OBJ_ID SRC_FILE BLOCK_SIZE BLOCK_COUNT BLOCKS_PER_IO UPDATE_FLAG OFFSET
read    OBJ_ID DEST_FILE BLOCK_SIZE BLOCK_COUNT BLOCKS_PER_IO
delete  OBJ_ID
help
quit
[root@ssc-vm-1623:root] consul kv get 0x7200000000000001:0x4
"0x7200000000000001:0x100000004"
[root@ssc-vm-1623:root] consul kv get 0x7200000000000001:0x5
"0x7200000000000001:0x300000005"
[root@ssc-vm-1623:root]
```

**Review by vaibhavparatwar — APPROVED** [review]

**Review by suvratjoshi — APPROVED** [review]

**Review by Shreya-18 — APPROVED** [review]

File: hax/hax/util.py [hunk 1]
```diff
@@ -1819,6 +1828,65 @@ def cleanup_node_process_states(self):
...
+                sleep(1)
+            self.process_dynamic_fidk_unlock()
+        return new_fidk
+
+    @repeat_if_fails()
+    def alloc_next_process_fid(self, process_fid: Fid) -> Fid:
+        next_fidk = self.get_process_next_dynamic_fidk_lock()
+        fid_mask: Fid = ObjTMaskMap[ObjT.PROCESS]
+        fid_cont = process_fid.container
+        fid_key = process_fid.key + ((fid_mask.key * next_fidk) + next_fidk)
```
Comment (Shreya-18) [inline hax/hax/util.py:1875]: @mssawant  Just a question: currently, we have assigned higher 32 bits as dynamic part. So later if we change the mask to lower 32, then this formula changes too, right? 
like in cfgen
`fidk = (((obj_mask.key ^ QW_F) * fidk + fidk) &
                obj_mask.key)`

Comment (mssawant) [reply hax/hax/util.py:1876]: That's the idea, formula should not change.
Please refer to https://seagate-systems.atlassian.net/wiki/spaces/PUB/pages/802258945/High+Level+Design+of+Dynamic+Motr+FID+Allocation+for+Motr+Client+processes#Partial-Fid-allocation, here I have illustrated the cases with reserving high and low 32 bits both. The formula does not change. Let me know if you have any doubts.

**Comment by mssawant** [comment]
Bootstrapped cluster,
```
[root@ssc-vm-1623:root] hctl status
Data pool:
    # fid name
    0x6f00000000000001:0x1 'sns-pool'
Profile:
    # fid name: pool(s)
    0x7000000000000001:0x0 'default': 'sns-pool' 'dix-pool' None
Services:
    ssc-vm-1623.colo.seagate.com  (RC)
    [started]  hax        0x7200000000000001:0x0   inet:tcp:10.230.250.51@2001
    [started]  confd      0x7200000000000001:0x1   inet:tcp:10.230.250.51@3001
    [started]  ioservice  0x7200000000000001:0x2   inet:tcp:10.230.250.51@3002
    [started]  ioservice  0x7200000000000001:0x3   inet:tcp:10.230.250.51@3003
    [started]  s3server   0x7200000000000001:0x4   inet:tcp:10.230.250.51@4001
    [offline]  m0_client  0x7200000000000001:0x5   inet:tcp:10.230.250.51@5001
    [unknown]  m0_client  0x7200000000000001:0x6   inet:tcp:10.230.250.51@5002
```

Started m0client and stopped and restarted.
```
[root@ssc-vm-1623:root] m0client -l 'inet:tcp:10.230.250.51@5001' -H 'inet:tcp:10.230.250.51@2001' -p '<0x7000000000000001:0x0>' -P '0x7200000000000001:0x5'
m0client >>quit
Help:
touch   OBJ_ID
write   OBJ_ID SRC_FILE BLOCK_SIZE BLOCK_COUNT BLOCKS_PER_IO UPDATE_FLAG OFFSET
read    OBJ_ID DEST_FILE BLOCK_SIZE BLOCK_COUNT BLOCKS_PER_IO
delete  OBJ_ID
help
quit
[root@ssc-vm-1623:root] m0client -l 'inet:tcp:10.230.250.51@5001' -H 'inet:tcp:10.230.250.51@2001' -p '<0x7000000000000001:0x0>' -P '0x7200000000000001:0x5'
m0client >>quit
```
```
201290  37209610.546.157514  5ff27b0  DEBUG   ha/note.c:225:m0_ha_msg_accept                      nv_note[0]=(no_id=<7200000000000001:200000005> no_state=2)
201291  37209610.546.158802  5ff26b0  CALL    ha/note.c:158:ha_state_accept                       > confc=0x7fff2debfa78 note->nv_nr=1
201292  37209610.546.167400  5ff26b0  DEBUG   ha/note.c:176:ha_state_accept                       nv_note[0]=(no_id=<7200000000000001:200000005> no_state=2) obj=(nil) obj->co_status=-1
201293  37209610.546.167804  5ff26b0  CALL    ha/note.c:186:ha_state_accept                       <
201294  37209610.546.168074  5ff26b0  CALL    ha/note.c:158:ha_state_accept                       > confc=0x7fff2debea38 note->nv_nr=1
201295  37209610.546.242724  5ff26b0  DEBUG   ha/note.c:176:ha_state_accept                       nv_note[0]=(no_id=<7200000000000001:200000005> no_state=2) obj=(nil) obj->co_status=-1
201296  37209610.546.243038  5ff26b0  CALL    ha/note.c:186:ha_state_accept                       <
```
updated m0client fid reported to motr ioservice, piggy backed over outgoing ha messages.

**Pull request merged**